File: packages/java/src/main/java/io/opensms/SenderIds.java
package io.opensms;

import com.fasterxml.jackson.databind.JsonNode;
import io.opensms.models.Page;
import io.opensms.models.SenderDocument;
import io.opensms.models.SenderId;
import io.opensms.models.SenderIdCheck;
import io.opensms.models.SenderIdDraft;
import io.opensms.models.SenderIdQuote;

import java.util.Collections;
import java.util.List;

/** Sender ID registrations, drafts and documents. Accessed as {@code client.senderIds()}. */
public final class SenderIds {

    private final ApiTransport t;

    SenderIds(ApiTransport t) {
        this.t = t;
    }

    /** @return the first page of sender IDs. */
    public Page<SenderId> list() {
        return list(new ListParams());
    }

    /**
     * @param params paging.
     * @return one page.
     */
    public Page<SenderId> list(ListParams params) {
        return t.get("/v1/sender-ids", params.toQuery(), ApiTransport.pageOf(SenderId.class));
    }

    /**
     * @param id sender ID id.
     * @return the sender ID with its registrations.
     */
    public SenderId get(String id) {
        return t.get("/v1/sender-ids/" + Wire.id(id, "id"), null, ApiTransport.type(SenderId.class));
    }

    /**
     * Register a sender ID. May charge a fee, so it is never retried automatically.
     *
     * @param params the registration.
     * @return the sender ID.
     */
    public SenderId create(SenderIdCreateParams params) {
        return t.json("POST", "/v1/sender-ids", params.toWire(), null, ApiTransport.type(SenderId.class));
    }

    /**
     * Amend a registration.
     *
     * @param id     sender ID id.
     * @param params the amendment.
     * @return the sender ID.
     */
    public SenderId update(String id, SenderIdUpdateParams params) {
        return t.json("PATCH", "/v1/sender-ids/" + Wire.id(id, "id"), params.toWire(), null,
                ApiTransport.type(SenderId.class));
    }

    /** @param id sender ID id. */
    public void delete(String id) {
        t.json("DELETE", "/v1/sender-ids/" + Wire.id(id, "id"), null, null, ApiTransport.VOID);
    }

    /**
     * @param value   the sender ID text.
     * @param country ISO2 country, or {@code null}.
     * @return {@code {valid, available, reserved, reason}}.
     */
    public SenderIdCheck check(String value, String country) {
        Wire.require(value, "value");
        return t.get("/v1/sender-ids/check", Query.of().add("value", value).add("country", country),
                ApiTransport.type(SenderIdCheck.class));
    }

    /**
     * @param countries ISO2 markets, sent as {@code countries=KE,NG}.
     * @return the fee quote.
     */
    public SenderIdQuote quote(List<String> countries) {
        if (countries == null || countries.isEmpty()) {
            throw new IllegalArgumentException("countries is required");
        }
        return t.get("/v1/sender-ids/quote", Query.of().add("countries", countries),
                ApiTransport.type(SenderIdQuote.class));
    }

    /** @return uploaded supporting documents (not paginated). */
    public List<SenderDocument> listDocuments() {
        JsonNode root = t.get("/v1/sender-documents", null, ApiTransport.type(JsonNode.class));
        if (root == null || root.get("items") == null || root.get("items").isNull()) {
            return Collections.emptyList();
        }
        return Json.MAPPER.convertValue(root.get("items"), ApiTransport.listOf(SenderDocument.class));
    }

    /** @return the first page of drafts. */
    public Page<SenderIdDraft> listDrafts() {
        return listDrafts(new ListParams());
    }

    /**
     * @param params paging.
     * @return one page.
     */
    public Page<SenderIdDraft> listDrafts(ListParams params) {
        return t.get("/v1/sender-id-drafts", params.toQuery(), ApiTransport.pageOf(SenderIdDraft.class));
    }

    /**
     * Create a draft. Not retried automatically (no idempotency support).
     *
     * @param params draft fields.
     * @return the draft ({@code version} 1).
     */
    public SenderIdDraft createDraft(SenderIdDraftParams params) {
        return t.json("POST", "/v1/sender-id-drafts", params.toWire(), null, ApiTransport.type(SenderIdDraft.class));
    }

    /**
     * @param id draft id.
     * @return the draft.
     */
    public SenderIdDraft getDraft(String id) {
        return t.get("/v1/sender-id-drafts/" + Wire.id(id, "id"), null, ApiTransport.type(SenderIdDraft.class));
    }

    /**
     * @param id     draft id.
     * @param params {@code version} (current) required, plus fields to change.
     * @return the draft with an incremented version.
     */
    public SenderIdDraft updateDraft(String id, SenderIdDraftParams params) {
        Wire.require(params.version(), "version");
        return t.json("PATCH", "/v1/sender-id-drafts/" + Wire.id(id, "id"), params.toWire(), null,
                ApiTransport.type(SenderIdDraft.class));
    }

    /** @param id draft id. */
    public void deleteDraft(String id) {
        t.json("DELETE", "/v1/sender-id-drafts/" + Wire.id(id, "id"), null, null, ApiTransport.VOID);
    }
}

File: packages/php/src/Resources/SenderIds.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `senderIds` resource: sender ID registrations, availability checks, fee quotes, documents and application drafts.
 * Accessed as `$client->senderIds`.
 */
final class SenderIds
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * @param array<string, mixed> $params `limit`, `cursor`
     */
    public function list(array $params = []): Page
    {
        return Models::page($this->transport->request('GET', '/v1/sender-ids', Models::map($params, Models::PAGE_QUERY, [], true)));
    }

    /**
     * @return array<string, mixed>
     */
    public function get(string $id): array
    {
        return Models::object($this->transport->request('GET', '/v1/sender-ids/' . Models::pathParam($id)));
    }

    /**
     * Register a sender ID. May charge fees (see {@see quote()}), so it is
     * never retried automatically. Params: `value`, `kind`, `countries`,
     * `documents` (required), `use_case`, `sample_message`, `draft_id`,
     * `draft_version`, `quote_id`.
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed>
     */
    public function create(array $params): array
    {
        $body = Models::map($params, Models::SENDER_ID_CREATE, ['value', 'kind', 'countries', 'documents']);

        return Models::object($this->transport->request('POST', '/v1/sender-ids', [], $body, ['retry' => false]));
    }

    /**
     * Amend a registration: `use_case`, `countries`, `documents`, `sample_message`.
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed>
     */
    public function update(string $id, array $params): array
    {
        $body = Models::map($params, Models::SENDER_ID_UPDATE);

        return Models::object($this->transport->request('PATCH', '/v1/sender-ids/' . Models::pathParam($id), [], $body));
    }

    public function delete(string $id): void
    {
        $this->transport->request('DELETE', '/v1/sender-ids/' . Models::pathParam($id));
    }

    /**
     * Check a value. Returns `{valid, available, reserved, reason}`.
     *
     * @param array{value: string, country?: string} $params
     * @return array<string, mixed>
     */
    public function check(array $params): array
    {
        $query = Models::map($params, Models::SENDER_ID_CHECK_QUERY, ['value'], true);

        return Models::object($this->transport->request('GET', '/v1/sender-ids/check', $query));
    }

    /**
     * Registration fee quote. `countries` is a list, sent as `KE,NG`.
     *
     * @param array{countries: list<string>|string} $params
     * @return array<string, mixed>
     */
    public function quote(array $params): array
    {
        $query = Models::map($params, Models::SENDER_ID_QUOTE_QUERY, ['countries'], true);

        return Models::object($this->transport->request('GET', '/v1/sender-ids/quote', $query));
    }

    /**
     * Uploaded registration documents (upload and download are console only).
     *
     * @return list<array<string, mixed>>
     */
    public function listDocuments(): array
    {
        $response = Models::object($this->transport->request('GET', '/v1/sender-documents'));

        return Models::list($response['items'] ?? []);
    }

    /**
     * @param array<string, mixed> $params `limit`, `cursor`
     */
    public function listDrafts(array $params = []): Page
    {
        return Models::page($this->transport->request('GET', '/v1/sender-id-drafts', Models::map($params, Models::PAGE_QUERY, [], true)));
    }

    /**
     * Params: `source`, `value`, `kind`, `countries`, `use_case`,
     * `sample_message`, `documents`. Not retried automatically.
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed>
     */
    public function createDraft(array $params = []): array
    {
        $body = Models::map($params, Models::SENDER_ID_DRAFT);

        return Models::object($this->transport->request('POST', '/v1/sender-id-drafts', [], $body, ['retry' => false]));
    }

    /**
     * @return array<string, mixed>
     */
    public function getDraft(string $id): array
    {
        return Models::object($this->transport->request('GET', '/v1/sender-id-drafts/' . Models::pathParam($id)));
    }

    /**
     * Update a draft. `version` (the current version) is required; a stale
     * version returns 409.
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed>
     */
    public function updateDraft(string $id, array $params): array
    {
        $body = Models::map($params, Models::SENDER_ID_DRAFT_UPDATE, ['version']);

        return Models::object($this->transport->request('PATCH', '/v1/sender-id-drafts/' . Models::pathParam($id), [], $body));
    }

    public function deleteDraft(string $id): void
    {
        $this->transport->request('DELETE', '/v1/sender-id-drafts/' . Models::pathParam($id));
    }
}
